src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProject.cs

```diff
@@ -1231,7 +1231,17 @@ public void RemoveFromWorkspace()
                 // references being converted to metadata references (or vice versa) and we might either
                 // miss stopping a file watcher or might end up double-stopping a file watcher.
                 remainingMetadataReferences = w.CurrentSolution.GetRequiredProject(Id).MetadataReferences;
-                w.OnProjectRemoved(Id);
+                _workspace.RemoveProjectFromTrackingMaps_NoLock(Id);
+
+                // If this is our last project, clear the entire solution.
+                if (w.CurrentSolution.ProjectIds.Count == 1)
+                {
+                    _workspace.RemoveSolution_NoLock();
+                }
+                else
+                {
+                    _workspace.OnProjectRemoved(Id);
```

Comment:
I'm a little confused by the mixed usage of `_workspace` and `w` in the lambda. They seem to refer to the same object. If so, maybe use one of them consistently?

Reply:
They are the same object; the lambda parameter though is just typed as Workspace since I think my original intent was you don't call extra methods like this directly. But I think that was moot. I'll do a follow up PR for this.